### PR TITLE
just for getting feedback from mentor. I'm not sure that i'm moving i…

### PR DESCRIPTION
**Pull request by 163onmyneck**
…n a right way.

### CODE REVIEW

File: src/main/java/com/example/demo/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+public class ShoppingCartController {
+    private final UserService userService;
+    private final ShoppingCartService shoppingCartService;
+    private final CartItemService cartItemService;
+
+    @GetMapping
+    public ShoppingCartDto getShoppingCart(Authentication authentication) {
+        UserDetails userDetails = (UserDetails) authentication.getPrincipal();
+        Long currentUser = userService.getUserIdByEmail(userDetails.getUsername());
+        return shoppingCartService.getByUserId(currentUser);
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/controller/ShoppingCartController.java]: You can make only one call to DB having user Email. Just use SQL JOIN to have.

File: src/main/java/com/example/demo/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @PostMapping
+    public CartItemResponseDto addBook(@RequestBody CartItemRequestDto cartItemRequestDto,
+                                       Authentication authentication) {
+        return cartItemService.saveItemToShoppingCart(cartItemRequestDto,
+            getShoppingCart(authentication));
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/controller/ShoppingCartController.java]: In Spring Boot, it's generally not recommended for controller methods to directly call each other. Controller methods are typically designed to handle incoming HTTP requests and delegate processing to services or other components. Directly calling one controller method from another can lead to tight coupling between them, making the code harder to understand, test, and maintain.

Instead, it's better to extract common functionality into separate service classes and have the controller methods delegate to these services. This promotes separation of concerns and allows for better code organization and reusability. Additionally, services can be easily tested in isolation, which is crucial for maintaining code quality.

File: src/main/java/com/example/demo/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    }
+
+    @PutMapping("/cart-items/{cartItemId}")
+    public void updateQuantity(@PathVariable Long id, int quantity) {
+        cartItemService.getById(id).setQuantity(quantity);
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/controller/ShoppingCartController.java]: Controllers should not do any business logic, the just call proper services. setQuantity is the business logic.

File: src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public CartItemResponseDto getById(Long id) {
+        return cartItemMapper.toDto(cartItemRepository.findById(id).orElseThrow(
+            () -> new EntityNotFoundException("Can not find cart item by id:" + id)));
+    }
+
+    @Override
+    public void deleteById(Long id) {
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java:48]: put the @Transactional on this method. You don't save the entity

File: src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+                .save(cartItemMapper.toModel(cartItemRequestDto, bookRepository))));
+        return cartItemMapper.toDto(
+            cartItemRepository.save(cartItemMapper.toModel(cartItemRequestDto, bookRepository)));
+    }
+
+    @Override
+    public CartItemResponseDto getById(Long id) {
+        return cartItemMapper.toDto(cartItemRepository.findById(id).orElseThrow(
+            () -> new EntityNotFoundException("Can not find cart item by id:" + id)));
+    }
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java:44]: put the @Transactional on this method. You don't save the entity

File: src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+@Service
+@RequiredArgsConstructor
+public class CartItemServiceImpl implements CartItemService {
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public CartItemResponseDto saveItemToShoppingCart(CartItemRequestDto cartItemRequestDto,
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/service/cart/item/impl/CartItemServiceImpl.java:26]: @Transactional. You make more than one call to DB

File: src/main/java/com/example/demo/service/shopping/cart/ShoppingCartService.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package com.example.demo.service.shopping.cart;
+
+import com.example.demo.dto.shopping.cart.ShoppingCartDto;
+
+public interface ShoppingCartService {
+    ShoppingCartDto getByUserId(Long id);
+
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/service/shopping/cart/ShoppingCartService.java]:
```suggestion
```

File: src/main/java/com/example/demo/repository/shopping/cart/ShoppingCartRepository.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
+package com.example.demo.repository.shopping.cart;
+
+import com.example.demo.model.ShoppingCart;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.data.repository.query.Param;
+
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("FROM ShoppingCart sc WHERE sc.user.id = :id")
+    ShoppingCart findByUserId(@Param("id") Long id);
```
Comment (MAMentorFX) [inline src/main/java/com/example/demo/repository/shopping/cart/ShoppingCartRepository.java]: It should return Optional

**Review by MAMentorFX — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/mapper/BookMapper.java [hunk 9]
```diff
@@ -17,6 +20,11 @@ public interface BookMapper {
 
     BookDtoWithoutCategoryIds toDtoWithoutCategories(Book book);
 
+    @Named("bookFromId")
+    default Book bookFromId(Long id, @Context BookRepository bookRepository) {
```
Comment (Rommelua) [inline src/main/java/com/example/demo/mapper/BookMapper.java]: mapper layer can not use repository

File: src/main/java/com/example/demo/mapper/CartItemMapper.java [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+import org.mapstruct.Context;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class, uses = BookMapper.class)
+public interface CartItemMapper {
+    CartItemResponseDto toDto(CartItem cartItem);
+
+    @Mapping(target = "book", source = "bookId", qualifiedByName = "bookFromId")
+    CartItem toModel(CartItemRequestDto cartItemRequestDto, @Context BookRepository bookRepository);
```
Comment (Rommelua) [inline src/main/java/com/example/demo/mapper/CartItemMapper.java]: mapper layer can not use repository

File: src/main/java/com/example/demo/model/ShoppingCart.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+@Setter
+@SQLDelete(sql = "UPDATE shopping_carts SET is_deleted = true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Entity
+@Table(name = "shopping_carts")
+public class ShoppingCart {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne
```
Comment (Rommelua) [inline src/main/java/com/example/demo/model/ShoppingCart.java]: one to one

**Review by Rommelua — CHANGES_REQUESTED** [review]

**Review by koval-kn — APPROVED** [review]

**Pull request merged**